blog/2020/05/rr.md

```diff
@@ -0,0 +1,452 @@
+enforcement of safety for regions where this cannot be proven by the compiler.
+However, that is for future work. For the moment, recording multi-threaded,
+shared-memory applications will likely incur overhead linear in the number of
+threads. It is thus a good idea to try and reproduce any issues at low core count.
+
+With that caveat in mind. Here is some real benchmark numbers of overhead for
+the julia test suite. The tests suite mostly uses message-passing based
+parallelism for running multiple tests in parallel. Only the `threads` test
+incurs the shared memory overhead penalty. Here are the measurements:
+
+[data]
+
+The total compressed size of a trace for a full run of the test suite (about
+30 minutes on 10 cores) is about three gigabytes.
+
+# Hardware and Software limitations
+
+As was already mentioned, `rr` currently only works on Linux. However, there
+are more restrictions. At the moment only x86 chips with Intel microarchitecture
+are supported. `rr` relies on hardware performance counters to establish a
+precise and accurate notion of time for asynchronous events. If these hardware
+counters are not available or not precise enough, recording using `rr` will not
+work. We and others have investigated whether it would be possible to port rr
+to other architectures. To the best of my knowledge, the current thinking on
+this is as follows:
+
+- For x86_64 with AMD Ryzen microarchitecture, it seems potentially possible, but the
+    AMD performance counters are less accurate than those found on Intel chips.
+    It may be possible to compensate for this in software, but the in-accuracies
+    are ill-understood. Help in investigation is welcome ([issue](https://github.com/mozilla/rr/issues/2034)).
+
+- For Aarch64, ll/sc-based atomics introduce additional problems, since various
+    external factors (interrupts, descheduling, etc.) can be observed as sc aborts.
+    With proper hardware support, this would be possible to work around, but such
+    hardware support appeared to be too in-accurate on older generation Aarch64 chips
+    and have been removed on recent generation ones. That said, Aarch64
+    microarchitectures are varied and as evidenced by the x86_64 experience, these
+    things can vary quite a bit by microarchitecture. Investigation of the hardware
+    capabilities of various Aarch64 microarchitectures would be very welcome
+    ([issue](https://github.com/mozilla/rr/issues/1373)).
+
+- We have looked into rr on POWER9, which is found in several large supercomputers
+    that run julia jobs. POWER9 has similar issues to Aarch64 since it is also an
+    ll/sc architecture, as well as supporting hardware transactional memory. In theory,
+    we believe the hardware exists to compensate for these issues, but preliminary
+    investigation suggests the hardware is not accurate enough to support `rr`. That
+    said, I understand there's active discussions with IBM to determine whether this
```

Comment:
How strongly do you want to stand behind this rumor?

Reply:
Strongly enough such that if anybody from the IBM hardware side reads this, they'll know to ask around internally and find the right people to get in touch with us.